pep-0692.rst

```diff
@@ -0,0 +1,599 @@
+PEP: 692
+Title: Using TypedDict for more precise \*\*kwargs typing
+Author: Franek Magiera <framagie@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-May-2022
+Python-Version: 3.12
+Post-History: 29-May-2022
+
+
+Abstract
+========
+
+Currently ``**kwargs`` can be type hinted as long as all of the keyword
+arguments specified by them are of the same type. However, that behaviour can
+be very limiting. Therefore, in this PEP we propose a new way to enable more
+precise ``**kwargs`` typing. The new approach revolves around using
+``TypedDict`` to type ``**kwargs`` that comprise keyword arguments of different
+types. It also involves introducing a grammar change and a new dunder
+``__unpack__``\.
+
+Motivation
+==========
+
+Currently annotating ``**kwargs`` with a type ``T`` means that the ``kwargs``
+type is in fact ``dict[str, T]``. For example:
+
+::
+
+    def foo(**kwargs: str) -> None: ...
+
+means that all keyword arguments in ``foo`` are strings (i.e., ``kwargs`` is
+of type ``dict[str, str]``). This behaviour limits the ability to type
+annotate ``**kwargs`` only to the cases where all of them are of the same type.
+However, it is often the case that keyword arguments conveyed by ``**kwargs``
+have different types that are dependent on the keyword's name. In those cases
+type annotating ``**kwargs`` is not possible. This is especially a problem for
+already existing codebases where the need of refactoring the code in order to
```

Comment:
Based on the conversation from “ Support for default types in TypedDict #1165”, I’m disinclined to believe the additional type hinting provided would be of benefit. The kwargs are hidden (abstracted) within a TypedDict, so the type hinting provided (esp. by certain language servers) doesn't provide useful information to the caller on what arguments to pass.

I recommend a note be added that this PEP is born out of convenience, not necessity, and that it be strongly encouraged wherever possible that the full argument list be type hinted.

Reply:
Thanks for bringing that up!

>I’m disinclined to believe the additional type hinting provided would be of benefit.

In my opinion the proposal could bring benefits, although I agree in some cases (new codebases where the keyword arguments are known from the start/typing via stubs) using "explicit" signatures is a better tool for the job.

>The kwargs are hidden (abstracted) within a TypedDict, so the type hinting provided (esp. by certain language servers) doesn't provide useful information to the caller on what arguments to pass.

Proposed type hinting would show the caller exactly what kind of keyword arguments the function expects. It is just the case of looking up how the TypedDict is defined (not an expert on language servers but seems like those could deal with "unpacking" the kwargs typed this way and provide a more user friendly hint). As I've mentioned in the response to the thread you linked, realistically not everyone will want to refactor and in that case I'd much rather see `**kwargs: **SomeTypedDictIHaveToLookUp` than `**kwargs: Any`. Also then, `**kwargs` could be treated as a `TypedDict` and type checked inside the function itself without the need of refactoring.

I think the discussion uncovered some valuable points that are worth adding to the draft (when it is better to use something else than what is being proposed). Let me try to include that.